ttnctl/cmd/downlink.go

```diff
@@ -29,30 +36,71 @@ var downlinkCmd = &cobra.Command{
 		}
 		ctx = ctx.WithField("DevID", devID)
 
-		payload, err := types.ParseHEX(args[1], len(args[1])/2)
+		JSON, err := cmd.Flags().GetBool("json")
+
 		if err != nil {
-			ctx.WithError(err).Fatal("Invalid Payload")
+			ctx.WithError(err).Fatal("Error: value of the json flag")
 		}
 
-		fPort, _ := cmd.Flags().GetInt("fport")
-
-		token := client.PublishDownlink(mqtt.DownlinkMessage{
-			AppID:   appID,
-			DevID:   devID,
-			FPort:   uint8(fPort),
-			Payload: payload,
-		})
-		token.Wait()
-		if token.Error() != nil {
-			ctx.WithError(token.Error()).Fatal("Could not enqueue downlink")
-		}
+		if JSON {
+			fields := args[1]
+			if fields == "" {
+				ctx.WithError(err).Fatal("No fields or payload provided")
+			}
+
+			// Valid payload provided + json flag
+			_, err := types.ParseHEX(args[1], len(args[1])/2)
+			if err == nil {
+				ctx.WithError(err).Fatal("You are providing a valid payload using the --json flag.")
+			}
+
+			fPort, _ := cmd.Flags().GetInt("fport")
 
-		ctx.Info("Enqueued downlink")
+			token := client.PublishDownlink(mqtt.DownlinkMessage{
+				AppID:  appID,
+				DevID:  devID,
+				FPort:  uint8(fPort),
+				Fields: parseJSON(fields),
+			})
+			token.Wait()
+			if token.Error() != nil {
+				ctx.WithError(token.Error()).Fatal("Could not enqueue downlink")
+			}
 
+			ctx.Info("Enqueued downlink")
+		} else { // Payload provided
+			payload, err := types.ParseHEX(args[1], len(args[1])/2)
+			if err != nil {
+				ctx.WithError(err).Fatal("Invalid Payload")
+			}
+
+			fPort, _ := cmd.Flags().GetInt("fport")
+
+			token := client.PublishDownlink(mqtt.DownlinkMessage{
+				AppID:   appID,
+				DevID:   devID,
+				FPort:   uint8(fPort),
+				Payload: payload,
+			})
+			token.Wait()
+			if token.Error() != nil {
+				ctx.WithError(token.Error()).Fatal("Could not enqueue downlink")
+			}
+
+			ctx.Info("Enqueued downlink")
```

Comment:
Do not repeat this. Set either the `Fields` or the `Payload` in a `mqtt.DownlinkMessage` depending on the json flag, and publish it once.

- Read FPort
- Read JSON flag
- Create `mqttDownlinkMessage`
- If JSON: parse it and set Fields, if not JSON: parse hex and set Payload
- Publish